### PR TITLE
ranking: Do incremental vacuuming, not all-or-nothing deletes

### PR DESCRIPTION
**Pull request by efritz**
In spirit of #48173. Previously, we were finding and deleting ALL records that are no longer relevant as a background GC process. This actually creates consistent load on the database since we have to scan a large number of records to find out we have nothing to delete.

The new query makes two behavioral changes from the old:

- We use a new timestamp _last scanned_ column to round-robin the records.
- We do not re-check the same rows within some configurable time frame. That's neurotic :)

## Test plan

Existing unit tests.

### CODE REVIEW

File: enterprise/internal/codeintel/uploads/internal/store/store_ranking.go [hunk 1]
```diff
@@ -354,6 +355,12 @@ SELECT
 	(SELECT COUNT(*) FROM inserted) AS num_inserted
 `
 
+// TODO - configure via envvar
+const vacuumBatchSize = 1000
+
+// TODO - configure via envvar
+var threshold = time.Duration(1) * time.Hour
```
Comment (eseliger) [inline enterprise/internal/codeintel/uploads/internal/store/store_ranking.go:362]: before merge?

File: enterprise/internal/codeintel/uploads/internal/store/store_ranking.go [hunk 2]
```diff
@@ -354,6 +355,12 @@ SELECT
 	(SELECT COUNT(*) FROM inserted) AS num_inserted
 `
 
+// TODO - configure via envvar
+const vacuumBatchSize = 1000
+
+// TODO - configure via envvar
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/store/store_ranking.go:361]: Can do this post-merge of this PR.

File: enterprise/internal/codeintel/uploads/internal/store/store_ranking.go [hunk 3]
```diff
@@ -383,31 +394,49 @@ func (s *store) VacuumStaleDefinitionsAndReferences(ctx context.Context, graphKe
 const vacuumStaleDefinitionsAndReferencesQuery = `
```
Comment (eseliger) [inline enterprise/internal/codeintel/uploads/internal/store/store_ranking.go:394]: can we get a query plan? 😬

**Review by eseliger — APPROVED** [review]

File: enterprise/internal/codeintel/uploads/internal/store/store_ranking.go (hunk 3 above)
Comment (efritz) [reply enterprise/internal/codeintel/uploads/internal/store/store_ranking.go:394]: **Before**:

```
                                                                                                         QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Result  (cost=4089203.59..4089203.60 rows=1 width=16) (actual time=24978.573..24978.586 rows=1 loops=1)
   CTE locked_definitions
     ->  LockRows  (cost=30552.64..2802922.73 rows=23970403 width=14) (actual time=2586.426..23500.139 rows=1361 loops=1)
           ->  Index Scan using codeintel_ranking_definitions_pkey on codeintel_ranking_definitions  (cost=30552.64..2563218.70 rows=23970403 width=14) (actual time=2586.408..23499.007 rows=1361 loops=1)
                 Filter: ((NOT (hashed SubPlan 1)) AND (graph_key = 'dotcom-test'::text))
                 Rows Removed by Filter: 47891184
                 SubPlan 1
                   ->  Index Only Scan using lsif_uploads_visible_at_tip_is_default_branch on lsif_uploads_visible_at_tip uvt  (cost=0.42..28541.42 rows=804263 width=4) (actual time=0.028..303.074 rows=820777 loops=1)
                         Heap Fetches: 361458
   CTE locked_references
     ->  LockRows  (cost=30552.50..182526.29 rows=536622 width=14) (actual time=718.050..1469.911 rows=24 loops=1)
           ->  Index Scan using codeintel_ranking_references_pkey on codeintel_ranking_references  (cost=30552.50..177160.07 rows=536622 width=14) (actual time=718.038..1469.803 rows=24 loops=1)
                 Filter: ((NOT (hashed SubPlan 3)) AND (graph_key = 'dotcom-test'::text))
                 Rows Removed by Filter: 1072597
                 SubPlan 3
                   ->  Index Only Scan using lsif_uploads_visible_at_tip_is_default_branch on lsif_uploads_visible_at_tip uvt_1  (cost=0.42..28541.42 rows=804263 width=4) (actual time=0.036..297.649 rows=820777 loops=1)
                         Heap Fetches: 361531
   CTE deleted_definitions
     ->  Delete on codeintel_ranking_definitions codeintel_ranking_definitions_1  (cost=539334.63..539819.32 rows=23970403 width=38) (actual time=23501.412..23506.858 rows=1361 loops=1)
           ->  Nested Loop  (cost=539334.63..539819.32 rows=23970403 width=38) (actual time=23501.398..23505.498 rows=1361 loops=1)
                 ->  HashAggregate  (cost=539334.07..539336.07 rows=200 width=40) (actual time=23501.368..23501.800 rows=1361 loops=1)
                       Group Key: locked_definitions.id
                       ->  CTE Scan on locked_definitions  (cost=0.00..479408.06 rows=23970403 width=40) (actual time=2586.437..23500.835 rows=1361 loops=1)
                 ->  Index Scan using codeintel_ranking_definitions_pkey on codeintel_ranking_definitions codeintel_ranking_definitions_1  (cost=0.56..2.78 rows=1 width=14) (actual time=0.002..0.002 rows=1 loops=1361)
                       Index Cond: (id = locked_definitions.id)
   CTE deleted_references
     ->  Delete on codeintel_ranking_references codeintel_ranking_references_1  (cost=12074.42..12527.16 rows=536622 width=38) (actual time=1470.936..1471.128 rows=24 loops=1)
           ->  Nested Loop  (cost=12074.42..12527.16 rows=536622 width=38) (actual time=1469.976..1470.055 rows=24 loops=1)
                 ->  HashAggregate  (cost=12074.00..12076.00 rows=200 width=40) (actual time=1469.953..1469.964 rows=24 loops=1)
                       Group Key: locked_references.id
                       ->  CTE Scan on locked_references  (cost=0.00..10732.44 rows=536622 width=40) (actual time=718.060..1469.934 rows=24 loops=1)
                 ->  Index Scan using codeintel_ranking_references_pkey on codeintel_ranking_references codeintel_ranking_references_1  (cost=0.43..2.62 rows=1 width=14) (actual time=0.003..0.003 rows=1 loops=24)
                       Index Cond: (id = locked_references.id)
   InitPlan 7 (returns $12)
     ->  Aggregate  (cost=539334.07..539334.08 rows=1 width=8) (actual time=23507.423..23507.423 rows=1 loops=1)
           ->  CTE Scan on deleted_definitions  (cost=0.00..479408.06 rows=23970403 width=0) (actual time=23501.415..23507.268 rows=1361 loops=1)
   InitPlan 8 (returns $13)
     ->  Aggregate  (cost=12074.00..12074.01 rows=1 width=8) (actual time=1471.144..1471.145 rows=1 loops=1)
           ->  CTE Scan on deleted_references  (cost=0.00..10732.44 rows=536622 width=0) (actual time=1470.938..1471.137 rows=24 loops=1)
 Planning Time: 0.603 ms
 Trigger for constraint fk_codeintel_ranking_reference on codeintel_ranking_references: time=994.668 calls=24
 Execution Time: 25989.648 ms
(42 rows)

Time: 26011.726 ms (00:26.012)
```

**After**:

```
                                                                                                              QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Result  (cost=4324.13..4324.14 rows=1 width=16) (actual time=1653.109..1653.124 rows=1 loops=1)
   CTE locked_definitions
     ->  Limit  (cost=0.56..1011.33 rows=1000 width=27) (actual time=875.496..877.030 rows=1000 loops=1)
           ->  LockRows  (cost=0.56..48350138.23 rows=47835395 width=27) (actual time=875.495..876.919 rows=1000 loops=1)
                 ->  Index Scan using codeintel_ranking_definitions_graph_key_last_scanned_at_id on codeintel_ranking_definitions rd  (cost=0.56..47871784.28 rows=47835395 width=27) (actual time=875.469..876.238 rows=1000 loops=1)
                       Index Cond: (graph_key = 'dotcom-test'::text)
                       Filter: ((last_scanned_at IS NULL) OR ((now() - last_scanned_at) >= '04:00:00'::interval))
                       SubPlan 1
                         ->  Index Only Scan using lsif_uploads_visible_at_tip_is_default_branch on lsif_uploads_visible_at_tip uvt  (cost=0.42..3.51 rows=6 width=0) (never executed)
                               Index Cond: (upload_id = rd.upload_id)
                               Heap Fetches: 0
                       SubPlan 2
                         ->  Index Only Scan using lsif_uploads_visible_at_tip_is_default_branch on lsif_uploads_visible_at_tip uvt_1  (cost=0.42..28541.42 rows=804263 width=4) (actual time=0.093..314.068 rows=820755 loops=1)
                               Heap Fetches: 360530
   CTE updated_definitions
     ->  Update on codeintel_ranking_definitions  (cost=21.82..564.25 rows=500 width=219) (actual time=124.475..124.475 rows=0 loops=1)
           ->  Nested Loop  (cost=21.82..564.25 rows=500 width=219) (actual time=1.028..5.682 rows=1000 loops=1)
                 ->  HashAggregate  (cost=21.25..23.19 rows=194 width=40) (actual time=1.003..1.473 rows=1000 loops=1)
                       Group Key: ld.id
                       ->  CTE Scan on locked_definitions ld  (cost=0.00..20.00 rows=500 width=40) (actual time=0.008..0.392 rows=1000 loops=1)
                             Filter: safe
                 ->  Index Scan using codeintel_ranking_definitions_pkey on codeintel_ranking_definitions  (cost=0.56..2.78 rows=1 width=179) (actual time=0.004..0.004 rows=1 loops=1000)
                       Index Cond: (id = ld.id)
   CTE deleted_definitions
     ->  Delete on codeintel_ranking_definitions codeintel_ranking_definitions_1  (cost=21.82..563.00 rows=500 width=38) (actual time=877.620..877.622 rows=0 loops=1)
           ->  Nested Loop  (cost=21.82..563.00 rows=500 width=38) (actual time=877.619..877.620 rows=0 loops=1)
                 ->  HashAggregate  (cost=21.25..23.19 rows=194 width=40) (actual time=877.618..877.619 rows=0 loops=1)
                       Group Key: ld_1.id
                       ->  CTE Scan on locked_definitions ld_1  (cost=0.00..20.00 rows=500 width=40) (actual time=877.616..877.616 rows=0 loops=1)
                             Filter: (NOT safe)
                             Rows Removed by Filter: 1000
                 ->  Index Scan using codeintel_ranking_definitions_pkey on codeintel_ranking_definitions codeintel_ranking_definitions_1  (cost=0.56..2.78 rows=1 width=14) (never executed)
                       Index Cond: (id = ld_1.id)
   CTE locked_references
     ->  Limit  (cost=0.43..1097.96 rows=1000 width=27) (actual time=772.305..774.936 rows=1000 loops=1)
           ->  LockRows  (cost=0.43..1174764.27 rows=1070373 width=27) (actual time=772.304..774.816 rows=1000 loops=1)
                 ->  Index Scan using codeintel_ranking_references_graph_key_last_scanned_at_id on codeintel_ranking_references rr  (cost=0.43..1164060.54 rows=1070373 width=27) (actual time=772.289..773.356 rows=1000 loops=1)
                       Index Cond: (graph_key = 'dotcom-test'::text)
                       Filter: ((last_scanned_at IS NULL) OR ((now() - last_scanned_at) >= '04:00:00'::interval))
                       SubPlan 6
                         ->  Index Only Scan using lsif_uploads_visible_at_tip_is_default_branch on lsif_uploads_visible_at_tip uvt_2  (cost=0.42..3.51 rows=6 width=0) (never executed)
                               Index Cond: (upload_id = rr.upload_id)
                               Heap Fetches: 0
                       SubPlan 7
                         ->  Index Only Scan using lsif_uploads_visible_at_tip_is_default_branch on lsif_uploads_visible_at_tip uvt_3  (cost=0.42..28541.42 rows=804263 width=4) (actual time=0.031..305.393 rows=820755 loops=1)
                               Heap Fetches: 360530
   CTE updated_references
     ->  Update on codeintel_ranking_references  (cost=21.68..533.17 rows=500 width=827) (actual time=57.492..57.492 rows=0 loops=1)
           ->  Nested Loop  (cost=21.68..533.17 rows=500 width=827) (actual time=0.986..5.058 rows=1000 loops=1)
                 ->  HashAggregate  (cost=21.25..23.19 rows=194 width=40) (actual time=0.961..1.405 rows=1000 loops=1)
                       Group Key: lr.id
                       ->  CTE Scan on locked_references lr  (cost=0.00..20.00 rows=500 width=40) (actual time=0.158..0.520 rows=1000 loops=1)
                             Filter: safe
                 ->  Index Scan using codeintel_ranking_references_pkey on codeintel_ranking_references  (cost=0.43..2.62 rows=1 width=783) (actual time=0.003..0.003 rows=1 loops=1000)
                       Index Cond: (id = lr.id)
   CTE deleted_references
     ->  Delete on codeintel_ranking_references codeintel_ranking_references_1  (cost=21.68..531.92 rows=500 width=38) (actual time=775.476..775.478 rows=0 loops=1)
           ->  Nested Loop  (cost=21.68..531.92 rows=500 width=38) (actual time=775.474..775.475 rows=0 loops=1)
                 ->  HashAggregate  (cost=21.25..23.19 rows=194 width=40) (actual time=775.473..775.473 rows=0 loops=1)
                       Group Key: lr_1.id
                       ->  CTE Scan on locked_references lr_1  (cost=0.00..20.00 rows=500 width=40) (actual time=775.470..775.471 rows=0 loops=1)
                             Filter: (NOT safe)
                             Rows Removed by Filter: 1000
                 ->  Index Scan using codeintel_ranking_references_pkey on codeintel_ranking_references codeintel_ranking_references_1  (cost=0.43..2.62 rows=1 width=14) (never executed)
                       Index Cond: (id = lr_1.id)
   InitPlan 11 (returns $20)
     ->  Aggregate  (cost=11.25..11.26 rows=1 width=8) (actual time=877.624..877.624 rows=1 loops=1)
           ->  CTE Scan on deleted_definitions  (cost=0.00..10.00 rows=500 width=0) (actual time=877.621..877.621 rows=0 loops=1)
   InitPlan 12 (returns $21)
     ->  Aggregate  (cost=11.25..11.26 rows=1 width=8) (actual time=775.480..775.480 rows=1 loops=1)
           ->  CTE Scan on deleted_references  (cost=0.00..10.00 rows=500 width=0) (actual time=775.477..775.477 rows=0 loops=1)
 Planning Time: 3.433 ms
 Execution Time: 1850.660 ms
```

**Pull request merged**